sdk/program/src/message/v0.rs

```diff
@@ -0,0 +1,385 @@
+    #[serde(with = "short_vec")]
+    pub address_maps: Vec<AddressMap>,
+
+    /// The blockhash of a recent block.
+    pub recent_blockhash: Hash,
+
+    /// Instructions that invoke a designated program, are executed in sequence,
+    /// and committed in one atomic transaction if all succeed.
+    ///
+    /// # Notes
+    ///
+    /// Account indices will index into the list of addresses constructed from
+    /// the concatenation of `account_keys` and the `entries` of each address
+    /// map in sequential order.
+    #[serde(with = "short_vec")]
+    pub instructions: Vec<CompiledInstruction>,
+}
+
+impl Sanitize for Message {
+    fn sanitize(&self) -> Result<(), SanitizeError> {
+        // signing area and read-only non-signing area should not
+        // overlap
+        if self.header.num_required_signatures as usize
+            + self.header.num_readonly_unsigned_accounts as usize
+            > self.account_keys.len()
+        {
+            return Err(SanitizeError::IndexOutOfBounds);
+        }
+
+        // there should be at least 1 RW fee-payer account.
+        if self.header.num_readonly_signed_accounts >= self.header.num_required_signatures {
+            return Err(SanitizeError::IndexOutOfBounds);
+        }
+
+        // there cannot be more address maps than read-only unsigned accounts.
+        if self.address_maps.len() > self.header.num_readonly_unsigned_accounts as usize {
+            return Err(SanitizeError::IndexOutOfBounds);
+        }
+
+        // each map must load at least one entry
+        let mut num_loaded_accounts = self.account_keys.len();
+        for map in &self.address_maps {
+            let num_loaded_map_entries = map
+                .read_only_entries
+                .len()
+                .saturating_add(map.writable_entries.len());
+
+            if num_loaded_map_entries == 0 {
+                return Err(SanitizeError::InvalidValue);
+            }
+
+            num_loaded_accounts = num_loaded_accounts.saturating_add(num_loaded_map_entries);
+        }
+
+        // the number of loaded accounts must be <= 256 since account indices are
+        // encoded as `u8`
+        if num_loaded_accounts > 256 {
+            return Err(SanitizeError::IndexOutOfBounds);
+        }
+
+        for ci in &self.instructions {
+            if ci.program_id_index as usize >= num_loaded_accounts {
+                return Err(SanitizeError::IndexOutOfBounds);
+            }
+            // A program cannot be a payer.
+            if ci.program_id_index == 0 {
+                return Err(SanitizeError::IndexOutOfBounds);
+            }
+            for ai in &ci.accounts {
+                if *ai as usize >= num_loaded_accounts {
+                    return Err(SanitizeError::IndexOutOfBounds);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Address map specifies which entries to load and
+#[frozen_abi(digest = "Bus4bNKecGtgSixqPn3vaaFmzyPVfnmLUwh6as9cZmQa")]
```

Comment:
this frozen_abi too: https://github.com/solana-labs/solana/pull/18725/files#r672290250